#402 - Unit rating performance fix

NickAragua opened the pull request:
This change prevents the re-initialization and subsequent re-evaluation (in about 99% of the cases) of the unit rating mechanism. This drastically improves performance for large save games and also cuts down the clutter in the log file. 

We still re-calculate the unit rating every day, whenever we do something that affects the unit rating, etc, as well as whenever we re-initialize the campaign object for whatever reason. However, we only do it once, as opposed to 200 times. :)

neoancient commented:
I've requested review from @dericpage because he has done a lot of work on the unit ratings code.

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8541,4 +8542,22 @@ public boolean checkYearlyRetirements() {
+		if(unitRating == null) {
+			UnitRatingMethod method = getCampaignOptions().getUnitRatingMethod();
+			
+		    if (method.equals(UnitRatingMethod.FLD_MAN_MERCS_REV)) {
```
This should be flipped around against the off-chance that _method_ somehow ends up being null.

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8541,4 +8542,22 @@ public boolean checkYearlyRetirements() {
+
+	/**
+	 * Returns the type of rating method as selected in the Campaign Options dialog. Lazy-loaded for performance.
+	 */
```
JavaDoc should probably mention that the default return value is a CampaignOpsReputation object.

dericpage commented on MekHQ/src/mekhq/gui/CampaignGUI.java:
```diff
@@ -2821,8 +2821,11 @@ private void refreshFunds() {
+            String text = "<html><b>Dragoons Rating:</b> " + getCampaign().getUnitRatingText() + "</html>";
             lblRating.setText(text);
         } else {
             lblRating.setText("");
```
I believe that when this method was written, FMMr was the only options for a unit rating, which is why the rating label only ever gets a value when that is selected.  It seems that, now that we have COR as an alternative, we probably should put in an _else if_ condition to populate the label with it's unitRatingText value.

NickAragua commented:
Added text update. Also fixed a bug where the unit rating method wouldn't change if you switched unit rating methods on the campaign options interface.

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8682,4 +8689,90 @@ public IUnitRating getUnitRating() {
+        return cost;
+    }
+
+    public long getMonthlySpareParts() {
```
Can be private.

pheonixstorm replied:
That was part of the peacetime costs feature

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8682,4 +8689,90 @@ public IUnitRating getUnitRating() {
+        return partsCost;
+    }
+
+    public long getMonthlyFuel() {
```
Can be private.

pheonixstorm replied:
Same

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8682,4 +8689,90 @@ public IUnitRating getUnitRating() {
+        return fuelCost;
+    }
+
+    public long getMonthlyAmmo() {
```
Can be private.

pheonixstorm replied:
Same

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8682,4 +8689,90 @@ public IUnitRating getUnitRating() {
+    }
+
+    public boolean isFactionComstar() {
+        if (getFactionCode().equals("CS")) {
```
This should be flipped to put the constant on the right in case getFactionCode returns null.

pheonixstorm replied:
Same

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8682,4 +8689,90 @@ public IUnitRating getUnitRating() {
+            case "CEI":
+            case "RA":
+                return true;
+            default: return false;
```
Should be on the following line.

dericpage commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -8682,4 +8689,90 @@ public IUnitRating getUnitRating() {
+    }
+
+    public boolean isFactionClan() {
+        switch (getFactionCode()) {
```
This has the potential to throw a NPE.  You should do a null check before the switch statement.

pheonixstorm replied:
Same, though honestly if faction returned null then there is an even bigger bug somewhere else. the faction code should never return null in this case. See line 340 of this file.

dericpage replied:
While the faction code is not supposed to return a null, it's possible that future code could change this.  In any case, it's a best practice to do a null-check before invoking a switch statement.  Better to detect and gracefully handle the null along with logging a sensible error than allowing the switch statement to blow up with a generic NPE.

NickAragua replied:
I decided to re-use the Faction.java mechanism for identifying whether a faction is clan, to save on duplicated logic. The campaign's current faction is always set to something by default (based on the campaign's constructor), so I don't think a null check is really worthwhile.

dericpage replied:
Unfortunately, setFactionCode doesn't contain null protection, so it is possible to set the factionCode null.  Technically, the better solution would, indeed, be to make sure that factionCode is only ever set through setFactionCode and add a null check there.

dericpage commented on MekHQ/src/mekhq/campaign/rating/AbstractUnitRating.java:
```diff
@@ -413,6 +413,8 @@ protected BigDecimal getNumberUnits() {
      */
     protected abstract int calculateUnitRatingScore();
 
+    public abstract UnitRatingMethod getUnitRatingMethod();
```
Since you have this in the interface, you shouldn't put it in the abstract unless you're adding logic to it.

NickAragua commented:
Made requested changes.
